src/groupeddatatable/grouping.jl

```diff
@@ -236,12 +236,28 @@ colwise(sum, groupby(dt, :a))
 ```
 
 """
-colwise(f::Function, d::AbstractDataTable) = map(i -> f(d[i]), 1:ncol(d))
+function colwise(f::Function, d::AbstractDataTable)
+    x = [f(d[i]) for i in 1:ncol(d)]
+    if eltype(x) <: Nullable && isa(x, Vector{eltype(x)})
+        return NullableArray(x)
+    end
+    return x
+end
 colwise(f::Function, gd::GroupedDataTable) = map(colwise(f), gd)
 colwise(f::Function) = x -> colwise(f, x)
 colwise(f) = x -> colwise(f, x)
 # apply several functions to each column in a DataTable
-colwise{T<:Function}(fns::Vector{T}, d::AbstractDataTable) = map(i -> map(f -> f(d[i]), fns), 1:ncol(d))
+function colwise{T<:Function}(fns::Vector{T}, d::AbstractDataTable)
+    res = Array{AbstractVector}(ncol(d))
```

Comment:
Let's go with a matrix then. Promotion doesn't sound like an issue to me: promotion rules should be precisely defined so that no information is lost (e.g. an integer can become a floating point but the new value is exactly equal to the original one).